### PR TITLE
Updating to typescript, adding token-to-regex functionality, and generating meta data on tokens

### PR DESCRIPTION
**Pull request by jeswr**
### Updating to typescript
All source files have been converted to typescript and relevant typescript definitions added

### token-to-regex functionaliy
The `reconstruct` function takes a tokenized expression and converts it back into a regular expression

### Generating metadata on tokens
The `addDetail` function creates additional metadata for tokenized expressions including
- The min length of a string that 'satisifies' a token (expression)
- The max length of a string that 'satisifies' a token (expression)
- The possible set of string that 'satisfy' a token (when finite)
- Regex validators for individual tokens (parts of a regular expression)
- Whether a token is satisfied by a single fixed value

(I've also updated and audited the dependencies - all tests are still passing)

### CODE REVIEW

**Comment by jeswr** [comment]
> This pull request **introduces 4 alerts** when merging [e74482a](https://github.com/fent/ret.js/commit/e74482ae4be47beaffbe7390660ef2898608d1ee) into [55250cc](https://github.com/fent/ret.js/commit/55250cce944784b407ca97e970dd0125fe17075a) - [view on LGTM.com](https://lgtm.com/projects/g/fent/ret.js/rev/pr-72e04b4d16603bd53d0aaf017977c5fd1fa13af3)
> 
> **new alerts:**
> 
>     * 4 for Syntax error

Note for whoever reviews the pull request - these alerts appear to be arising from the lgtm bot not being up to date with with typescript 4.0 or es2020 syntax.

**Comment by fent** [comment]
hi, thanks for the PR! I've been wanting to switch this project to ts, along with a few of my other projects. glad someone beat me to it :)

I'm going to start looking at this now, but I can say that typically, a PR introduces one change to the code. in this PR, 3 or 2 changes are introduced, changing to typescript, and adding `addDetail` and `reconstruct`. creating separate smaller PRs with separate concepts makes them easier to look at, not only for the reviewer, but for whoever is looking through commit or PR history in the future.

please separate this PR into two, one for typescript and one for the other changes. I'd argue that `addDetail` is added solely for `reconstruct`, so they can be in the same PR. if these changes are done on top of typescript changes, you can branch out from the typescript PR and apply those commits to the new feature branch.

File: lib/util.ts [hunk 1]
```diff
@@ -1,9 +1,9 @@
-const types = require('./types');
-const sets  = require('./sets');
+import { types, Range, Char, Set } from './types'
+import * as sets from './sets'
 
+type tokenClsArr = (Range | Char | Set)[]
```
Comment (fent) [inline lib/util.ts:4]: can be moved to `tokens.ts`, and could be named `SetTokens`

File: test/main-test.js [hunk 2]
```diff
@@ -450,7 +450,7 @@ vows.describe('Regexp Tokenizer')
       topic: ret('<(\\w+)>\\w*<\\1>'),
 
       'Reference a group': (t) => {
-        assert.deepEqual(t, {
+        assert.deepStrictEqual(t, {
```
Comment (fent) [inline test/main-test.js:453]: the `deepStrictEqual` changes could be kept to a separate PR to update tests

File: package.json [hunk 3]
```diff
@@ -21,17 +21,23 @@
     "typings"
   ],
```
Comment (fent) [inline package.json:22]: should change to just "dist" for `npm publish`

File: package.json [hunk 4]
```diff
@@ -21,17 +21,23 @@
...
   "scripts": {
-    "test": "nyc --extension --reporter=lcov --reporter=text-summary vows -- --spec test/*-test.js"
+    "build": "tsc",
```
Comment (fent) [inline package.json:24]: needs a `prepare` script that does the same thing as `build`

https://docs.npmjs.com/misc/scripts

File: tsconfig.json [hunk 5]
```diff
@@ -0,0 +1,69 @@
+{
+  "compilerOptions": {
+    /* Visit https://aka.ms/tsconfig.json to read more about this file */
+
+    /* Basic Options */
+    // "incremental": true,                   /* Enable incremental compilation */
+    "target": "es3",                          /* Specify ECMAScript target version: 'ES3' (default), 'ES5', 'ES2015', 'ES2016', 'ES2017', 'ES2018', 'ES2019', 'ES2020', or 'ESNEXT'. */
```
Comment (fent) [inline tsconfig.json:7]: "es6" is good

File: tsconfig.json [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+
+    /* Strict Type-Checking Options */
+    "strict": true,                           /* Enable all strict type-checking options. */
+    // "noImplicitAny": true,                 /* Raise error on expressions and declarations with an implied 'any' type. */
+    // "strictNullChecks": true,              /* Enable strict null checks. */
+    // "strictFunctionTypes": true,           /* Enable strict checking of function types. */
+    // "strictBindCallApply": true,           /* Enable strict 'bind', 'call', and 'apply' methods on functions. */
+    // "strictPropertyInitialization": true,  /* Enable strict checking of property initialization in classes. */
+    // "noImplicitThis": true,                /* Raise error on 'this' expressions with an implied 'any' type. */
+    // "alwaysStrict": true,                  /* Parse in strict mode and emit "use strict" for each source file. */
```
Comment (fent) [inline tsconfig.json:35]: i'd be ok with all these except `strickNullChecks` and `strickPropertyInitialization`

File: package.json [hunk 7]
```diff
@@ -21,17 +21,23 @@
...
+    "test": "tsc && nyc --extension --reporter=lcov --reporter=text-summary vows -- --spec test/*-test.js"
```
Comment (fent) [inline package.json:25]: would prefer to keep the test script separate from the build script. in case we're ever only working on tests, there'd be no need to recompile the main library

**Review by fent — CHANGES_REQUESTED** [review]
hi, i've added some comments.

also, what do you think about converting the tests too ts too? is that something you'd be interested in?

**Comment by jeswr** [comment]
> please separate this PR into two, one for typescript and one for the other changes.

Will do :)
> I'd argue that `addDetail` is added solely for `reconstruct`

`reconstruct` has no dependency on `addDetail`. `reconstruct` takes a `Root` token and returns the regular expression described by a token. `addDetail` takes a tokenized expression adds the following details (when applicable) to each token
```ts
type details = {
  minChar : number; // Minimum number of characters required to 'satisfy' the regex component
  maxChar : number; // Maximum number of characters required to 'satisfy' the regex component
  regexString: string; // Regular expression (string version) that the token is representing
  regex : RegExp; // Regular expression (Regex Object with flags) that the token is representing
  fixed: boolean; // Whether this token has a 'fixed solution', e.g `(hello)`, or many solutions, e.g. `[0-9]`.
  stringOptions: string[] | undefined; // List of options that satisfy the token
  reference?: number;
  leftEnd: boolean; // Whether the '^' positional is applied or not; 
  rightEnd: boolean; // Whether the '$' positional is applied or not;
} 
```
I will create 2 separate PRs; and also document these functions better in those PRs.

For context of why I created this - it is being used to generate customised form fields within an application I am working on.

**Comment by jeswr** [comment]
> hi, i've added some comments.

I'll go through those now
> also, what do you think about converting the tests too ts too? is that something you'd be interested in?

I'm pushed for time from now until mid November but happy to look into it after then if you would like.

File: lib/tokenizer.ts [hunk 8]
```diff
@@ -1,43 +1,44 @@
...
+  while (i < str.length) {
     c = str[i++];
 
     switch (c) {
       // Handle escaped characters, inclues a few sets.
       case '\\':
-        c = str[i++];
-
-        switch (c) {
+        switch (c = str[i++]) {
```
[the remark this answers is not in the document]
Comment (jeswr) [reply lib/tokenizer.ts:35]: I personally find it cleaner to read - happy to revert this if you disagree

File: lib/tokenizer.ts [hunk 9]
```diff
@@ -172,12 +171,16 @@ module.exports = (regexpStr) => {
...
         // Check if this group has a PIPE.
         // To get back the correct last stack.
-        last = lastGroup.options ?
-          lastGroup.options[lastGroup.options.length - 1] : lastGroup.stack;
+        const lastTemp = lastGroup.options ?
+        lastGroup.options[lastGroup.options.length - 1] : lastGroup.stack;
+      if (lastTemp)
+        last = lastTemp;
+      else
+        util.error(regexpStr, 'Invalid use of pipe');
```
[the remark this answers is not in the document]
Comment (jeswr) [reply lib/tokenizer.ts:183]: Pretty sure it was to keep Typescript happy when `strictNullChecks` where on. Happy to revert if you want to change this setting anyway

File: lib/tokenizer.ts [hunk 10]
```diff
@@ -186,13 +189,18 @@ module.exports = (regexpStr) => {
...
         }
-
         // Create a new stack and add to options for rest of clause.
-        let stack = [];
-        lastGroup.options.push(stack);
+        let stack: Token[] = [];
+        if (lastGroup.options)
+          lastGroup.options.push(stack);
+        else
+          util.error(regexpStr, 'last group does not contain options')
```
[the remark this answers is not in the document]
Comment (jeswr) [reply lib/tokenizer.ts:203]: Again, pretty sure it was to keep Typescript happy when strictNullChecks where on. Happy to revert if you want to change this setting anyway

File: lib/tokenizer.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (fent) [reply lib/tokenizer.ts:35]: either is fine. was wondering if the logic was different or there was some typescript quirk

File: lib/tokenizer.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (fent) [reply lib/tokenizer.ts:183]: ah, my worry is these lines not covered. not sure why the codecov check didn't run, I'll have to figure that out too.

**Comment by fent** [comment]
> I'm pushed for time from now until mid November but happy to look into it after then if you would like.

i can look into it after this PR, but you can beat me to it if i haven't done it by november
